src/esp/geo/VoxelGrid.cpp

```diff
@@ -0,0 +1,340 @@
+    positions.push_back(vertOffset + mid);
+    // Set the normals to be weighted such that cubes look slightly curved
+    normals.push_back(cubePositions[i].normalized() * 1 / 4 +
+                      cubeNormals[i].normalized() * 3 / 4);
+    colors.push_back(color);
+  }
+
+  for (int i = 0; i < 6; i++) {
+    if (!neighbors[i]) {
+      for (int j = 0; j < 6; j++) {
+        indices.push_back(sz + cubeIndices[i * 6 + j][0]);
+      }
+    }
+  }
+}
+
+void VoxelGrid::addVectorToMeshPrimitives(std::vector<Mn::Vector3>& positions,
+                                          std::vector<Mn::Vector3>& normals,
+                                          std::vector<Mn::Color3>& colors,
+                                          std::vector<Mn::UnsignedInt>& indices,
+                                          const Mn::Vector3i& local_coords,
+                                          const Mn::Vector3& vec) {
+  Mn::Trade::MeshData cubeData = Mn::Primitives::cubeSolid();
+
+  Mn::Vector3 mid = getGlobalCoords(local_coords);
+  Mn::Vector3 pos1 = vec.normalized() * m_voxelSize * 1 / 3 + mid;
+  Mn::Vector3 orthog1 = Mn::Math::cross(vec, Mn::Vector3(0, 1, 0));
+  if (orthog1 == Mn::Vector3(0, 0, 0)) {
+    orthog1 = Mn::Vector3(1, 0, 0);
+  }
+  Mn::Vector3 orthog2 = Mn::Math::cross(vec, orthog1);
+
+  Mn::Vector3 pos2 = mid + orthog1.normalized() * m_voxelSize * 1 / 15;
+  Mn::Vector3 pos3 = mid + orthog2.normalized() * m_voxelSize * 1 / 15;
+  Mn::Vector3 pos4 = mid - orthog1.normalized() * m_voxelSize * 1 / 15;
+  Mn::Vector3 pos5 = mid - orthog2.normalized() * m_voxelSize * 1 / 15;
+
+  unsigned int sz = positions.size();
+  positions.push_back(pos1);
+  positions.push_back(pos2);
+  positions.push_back(pos3);
+  positions.push_back(pos4);
+  positions.push_back(pos5);
+
+  colors.push_back(Mn::Color3(1, 1, 1));
+  colors.push_back(Mn::Color3(0, .3, 1));
+  colors.push_back(Mn::Color3(0, .3, 1));
+  colors.push_back(Mn::Color3(0, .3, 1));
+  colors.push_back(Mn::Color3(0, .3, 1));
+
+  normals.push_back(vec.normalized());
+  normals.push_back((pos1 - mid).normalized());
+  normals.push_back((pos2 - mid).normalized());
+  normals.push_back((pos3 - mid).normalized());
+  normals.push_back((pos4 - mid).normalized());
+
+  std::vector<unsigned int> inds{0, 1, 2, 0, 2, 3, 0, 3, 4,
+                                 0, 4, 1, 1, 3, 2, 1, 4, 3};
+  for (auto index : inds) {
+    indices.push_back(sz + index);
+  }
+
+  // Add mesh information for the stem of the arrow
+
+  // cube indices
+  sz = positions.size();
+  Mn::Vector3 pos6 = mid + orthog1.normalized() * m_voxelSize * 1 / 40;
+  Mn::Vector3 pos7 = mid + orthog2.normalized() * m_voxelSize * 1 / 40;
+  Mn::Vector3 pos8 = mid - orthog1.normalized() * m_voxelSize * 1 / 40;
+  Mn::Vector3 pos9 = mid - orthog2.normalized() * m_voxelSize * 1 / 40;
+
+  Mn::Vector3 pos10 = mid + orthog1.normalized() * m_voxelSize * 1 / 40 -
+                      vec.normalized() * m_voxelSize * 1 / 4;
```

Comment:
A lot of wasted cycles on renomralized ortho1, ortho2, and vec. Should just cache their normalized forms.